File: templates/py/mavlink_message.py
from __future__ import annotations

import struct
from abc import ABC, abstractmethod


class MavlinkMessage(ABC):
    @property
    @abstractmethod
    def mavlink_message_id(self) -> int:
        ...

    @property
    @abstractmethod
    def mavlink_crc_extra(self) -> int:
        ...

    @abstractmethod
    def serialize(self) -> bytes:
        ...

    @staticmethod
    def _endian() -> str:
        return "<"

    @classmethod
    def _get_int8(cls, data: bytes, offset: int) -> int:
        return struct.unpack_from(f"{cls._endian()}b", data, offset)[0]

    @classmethod
    def _get_uint8(cls, data: bytes, offset: int) -> int:
        return struct.unpack_from(f"{cls._endian()}B", data, offset)[0]

    @classmethod
    def _get_int16(cls, data: bytes, offset: int) -> int:
        return struct.unpack_from(f"{cls._endian()}h", data, offset)[0]

    @classmethod
    def _get_uint16(cls, data: bytes, offset: int) -> int:
        return struct.unpack_from(f"{cls._endian()}H", data, offset)[0]

    @classmethod
    def _get_int32(cls, data: bytes, offset: int) -> int:
        return struct.unpack_from(f"{cls._endian()}i", data, offset)[0]

    @classmethod
    def _get_uint32(cls, data: bytes, offset: int) -> int:
        return struct.unpack_from(f"{cls._endian()}I", data, offset)[0]

    @classmethod
    def _get_int64(cls, data: bytes, offset: int) -> int:
        return struct.unpack_from(f"{cls._endian()}q", data, offset)[0]

    @classmethod
    def _get_uint64(cls, data: bytes, offset: int) -> int:
        return struct.unpack_from(f"{cls._endian()}Q", data, offset)[0]

    @classmethod
    def _get_float32(cls, data: bytes, offset: int) -> float:
        return struct.unpack_from(f"{cls._endian()}f", data, offset)[0]

    @classmethod
    def _get_float64(cls, data: bytes, offset: int) -> float:
        return struct.unpack_from(f"{cls._endian()}d", data, offset)[0]

    @classmethod
    def _get_int8_list(cls, data: bytes, offset: int, length: int) -> list[int]:
        return [cls._get_int8(data, offset + i) for i in range(length)]

    @classmethod
    def _get_uint8_list(cls, data: bytes, offset: int, length: int) -> list[int]:
        return [cls._get_uint8(data, offset + i) for i in range(length)]

    @classmethod
    def _get_int16_list(cls, data: bytes, offset: int, length: int) -> list[int]:
        return [cls._get_int16(data, offset + i * 2) for i in range(length)]

    @classmethod
    def _get_uint16_list(cls, data: bytes, offset: int, length: int) -> list[int]:
        return [cls._get_uint16(data, offset + i * 2) for i in range(length)]

    @classmethod
    def _get_int32_list(cls, data: bytes, offset: int, length: int) -> list[int]:
        return [cls._get_int32(data, offset + i * 4) for i in range(length)]

    @classmethod
    def _get_uint32_list(cls, data: bytes, offset: int, length: int) -> list[int]:
        return [cls._get_uint32(data, offset + i * 4) for i in range(length)]

    @classmethod
    def _get_int64_list(cls, data: bytes, offset: int, length: int) -> list[int]:
        return [cls._get_int64(data, offset + i * 8) for i in range(length)]

    @classmethod
    def _get_uint64_list(cls, data: bytes, offset: int, length: int) -> list[int]:
        return [cls._get_uint64(data, offset + i * 8) for i in range(length)]

    @classmethod
    def _get_float32_list(cls, data: bytes, offset: int, length: int) -> list[float]:
        return [cls._get_float32(data, offset + i * 4) for i in range(length)]

    @classmethod
    def _get_float64_list(cls, data: bytes, offset: int, length: int) -> list[float]:
        return [cls._get_float64(data, offset + i * 8) for i in range(length)]

    @classmethod
    def _set_int8(cls, data: bytearray, offset: int, value: int) -> None:
        struct.pack_into(f"{cls._endian()}b", data, offset, value)

    @classmethod
    def _set_uint8(cls, data: bytearray, offset: int, value: int) -> None:
        struct.pack_into(f"{cls._endian()}B", data, offset, value)

    @classmethod
    def _set_int16(cls, data: bytearray, offset: int, value: int) -> None:
        struct.pack_into(f"{cls._endian()}h", data, offset, value)

    @classmethod
    def _set_uint16(cls, data: bytearray, offset: int, value: int) -> None:
        struct.pack_into(f"{cls._endian()}H", data, offset, value)

    @classmethod
    def _set_int32(cls, data: bytearray, offset: int, value: int) -> None:
        struct.pack_into(f"{cls._endian()}i", data, offset, value)

    @classmethod
    def _set_uint32(cls, data: bytearray, offset: int, value: int) -> None:
        struct.pack_into(f"{cls._endian()}I", data, offset, value)

    @classmethod
    def _set_int64(cls, data: bytearray, offset: int, value: int) -> None:
        struct.pack_into(f"{cls._endian()}q", data, offset, value)

    @classmethod
    def _set_uint64(cls, data: bytearray, offset: int, value: int) -> None:
        struct.pack_into(f"{cls._endian()}Q", data, offset, value)

    @classmethod
    def _set_float32(cls, data: bytearray, offset: int, value: float) -> None:
        struct.pack_into(f"{cls._endian()}f", data, offset, value)

    @classmethod
    def _set_float64(cls, data: bytearray, offset: int, value: float) -> None:
        struct.pack_into(f"{cls._endian()}d", data, offset, value)

    @classmethod
    def _set_int8_list(cls, data: bytearray, offset: int, values: list[int]) -> None:
        for i, value in enumerate(values):
            cls._set_int8(data, offset + i, value)

    @classmethod
    def _set_uint8_list(cls, data: bytearray, offset: int, values: list[int]) -> None:
        for i, value in enumerate(values):
            cls._set_uint8(data, offset + i, value)

    @classmethod
    def _set_int16_list(cls, data: bytearray, offset: int, values: list[int]) -> None:
        for i, value in enumerate(values):
            cls._set_int16(data, offset + i * 2, value)

    @classmethod
    def _set_uint16_list(cls, data: bytearray, offset: int, values: list[int]) -> None:
        for i, value in enumerate(values):
            cls._set_uint16(data, offset + i * 2, value)

    @classmethod
    def _set_int32_list(cls, data: bytearray, offset: int, values: list[int]) -> None:
        for i, value in enumerate(values):
            cls._set_int32(data, offset + i * 4, value)

    @classmethod
    def _set_uint32_list(cls, data: bytearray, offset: int, values: list[int]) -> None:
        for i, value in enumerate(values):
            cls._set_uint32(data, offset + i * 4, value)

    @classmethod
    def _set_int64_list(cls, data: bytearray, offset: int, values: list[int]) -> None:
        for i, value in enumerate(values):
            cls._set_int64(data, offset + i * 8, value)

    @classmethod
    def _set_uint64_list(cls, data: bytearray, offset: int, values: list[int]) -> None:
        for i, value in enumerate(values):
            cls._set_uint64(data, offset + i * 8, value)

    @classmethod
    def _set_float32_list(cls, data: bytearray, offset: int, values: list[float]) -> None:
        for i, value in enumerate(values):
            cls._set_float32(data, offset + i * 4, value)

    @classmethod
    def _set_float64_list(cls, data: bytearray, offset: int, values: list[float]) -> None:
        for i, value in enumerate(values):
            cls._set_float64(data, offset + i * 8, value)

File: templates/ts/mavlink_message.ts
export abstract class MavlinkMessage {
  abstract get mavlinkMessageId(): number;
  abstract get mavlinkCrcExtra(): number;
  abstract serialize(): Uint8Array;

  /** Message id check that survives duplicate ESM module instances. */
  static isMessageOf<T extends MavlinkMessage>(
    message: MavlinkMessage,
    messageClass: { readonly MSG_ID: number },
  ): message is T {
    return message.mavlinkMessageId === messageClass.MSG_ID;
  }

  private static view(data: Uint8Array): DataView {
    return new DataView(data.buffer, data.byteOffset, data.byteLength);
  }

  protected static getInt8(data: Uint8Array, offset: number): number {
    return MavlinkMessage.view(data).getInt8(offset);
  }

  protected static getUint8(data: Uint8Array, offset: number): number {
    return MavlinkMessage.view(data).getUint8(offset);
  }

  protected static getInt16(data: Uint8Array, offset: number): number {
    return MavlinkMessage.view(data).getInt16(offset, true);
  }

  protected static getUint16(data: Uint8Array, offset: number): number {
    return MavlinkMessage.view(data).getUint16(offset, true);
  }

  protected static getInt32(data: Uint8Array, offset: number): number {
    return MavlinkMessage.view(data).getInt32(offset, true);
  }

  protected static getUint32(data: Uint8Array, offset: number): number {
    return MavlinkMessage.view(data).getUint32(offset, true);
  }

  protected static getInt64(data: Uint8Array, offset: number): number {
    return Number(MavlinkMessage.view(data).getBigInt64(offset, true));
  }

  protected static getUint64(data: Uint8Array, offset: number): number {
    return Number(MavlinkMessage.view(data).getBigUint64(offset, true));
  }

  protected static getFloat32(data: Uint8Array, offset: number): number {
    return MavlinkMessage.view(data).getFloat32(offset, true);
  }

  protected static getFloat64(data: Uint8Array, offset: number): number {
    return MavlinkMessage.view(data).getFloat64(offset, true);
  }

  protected static getInt8List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) => MavlinkMessage.getInt8(data, offset + i));
  }

  protected static getUint8List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) => MavlinkMessage.getUint8(data, offset + i));
  }

  protected static getInt16List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) =>
      MavlinkMessage.getInt16(data, offset + i * 2),
    );
  }

  protected static getUint16List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) =>
      MavlinkMessage.getUint16(data, offset + i * 2),
    );
  }

  protected static getInt32List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) =>
      MavlinkMessage.getInt32(data, offset + i * 4),
    );
  }

  protected static getUint32List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) =>
      MavlinkMessage.getUint32(data, offset + i * 4),
    );
  }

  protected static getInt64List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) =>
      MavlinkMessage.getInt64(data, offset + i * 8),
    );
  }

  protected static getUint64List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) =>
      MavlinkMessage.getUint64(data, offset + i * 8),
    );
  }

  protected static getFloat32List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) =>
      MavlinkMessage.getFloat32(data, offset + i * 4),
    );
  }

  protected static getFloat64List(
    data: Uint8Array,
    offset: number,
    length: number,
  ): number[] {
    return Array.from({ length }, (_, i) =>
      MavlinkMessage.getFloat64(data, offset + i * 8),
    );
  }

  protected static setInt8(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setInt8(offset, value);
  }

  protected static setUint8(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setUint8(offset, value);
  }

  protected static setInt16(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setInt16(offset, value, true);
  }

  protected static setUint16(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setUint16(offset, value, true);
  }

  protected static setInt32(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setInt32(offset, value, true);
  }

  protected static setUint32(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setUint32(offset, value, true);
  }

  protected static setInt64(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setBigInt64(offset, BigInt(value), true);
  }

  protected static setUint64(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setBigUint64(offset, BigInt(value), true);
  }

  protected static setFloat32(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setFloat32(offset, value, true);
  }

  protected static setFloat64(data: Uint8Array, offset: number, value: number): void {
    MavlinkMessage.view(data).setFloat64(offset, value, true);
  }

  protected static setInt8List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setInt8(data, offset + i, values[i]!);
    }
  }

  protected static setUint8List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setUint8(data, offset + i, values[i]!);
    }
  }

  protected static setInt16List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setInt16(data, offset + i * 2, values[i]!);
    }
  }

  protected static setUint16List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setUint16(data, offset + i * 2, values[i]!);
    }
  }

  protected static setInt32List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setInt32(data, offset + i * 4, values[i]!);
    }
  }

  protected static setUint32List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setUint32(data, offset + i * 4, values[i]!);
    }
  }

  protected static setInt64List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setInt64(data, offset + i * 8, values[i]!);
    }
  }

  protected static setUint64List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setUint64(data, offset + i * 8, values[i]!);
    }
  }

  protected static setFloat32List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setFloat32(data, offset + i * 4, values[i]!);
    }
  }

  protected static setFloat64List(
    data: Uint8Array,
    offset: number,
    values: number[],
  ): void {
    for (let i = 0; i < values.length; i++) {
      MavlinkMessage.setFloat64(data, offset + i * 8, values[i]!);
    }
  }
}
